kyaml/setters2/settersutil/substitutioncreator.go

```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
+// values if they don't already exist in openAPIPath file.
+func (c SubstitutionCreator) CreateSettersForSubstitution(openAPIPath string) error {
+	b, err := ioutil.ReadFile(openAPIPath)
+	if err != nil {
+		return err
+	}
+
+	// parse the yaml file
+	y, err := yaml.Parse(string(b))
+	if err != nil {
+		return err
+	}
+
+	m, err := c.GetValuesForMarkers()
+	if err != nil {
+		return err
+	}
+
+	// for each ref in values, check if the setter already exists, if not create them
+	for _, value := range c.Values {
+		obj, err := y.Pipe(yaml.Lookup(
+			// get the setter key from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+			// extract io.k8s.cli.setters.image_setter
+			"openAPI", "definitions", strings.Split(value.Ref, "/")[2]))
+
+		if err != nil {
+			return err
+		}
+
+		if obj == nil {
+			sd := setters2.SetterDefinition{
+				// get the setter name from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+				// extract image_setter
+				Name:        strings.Split(value.Ref, ".")[4],
+				Value:       m[value.Marker],
+				SetBy:       "",
+				Description: "Created for substitution " + c.Name,
+				Count:       0,
+			}
+			sd.AddToFile(openAPIPath)
+		}
+	}
+	return nil
+}
+
+// GetValuesForMarkers parses the pattern and field value to derive values for the
+// markers in the pattern string. Returns error if the marker values can't be derived
+func (c SubstitutionCreator) GetValuesForMarkers() (map[string]string, error) {
+	m := make(map[string]string)
+	inds, _ := c.GetStartIndices()
+	s:=c.FieldValue
+	p:=c.Pattern
+	i:=0
+	j:=0
+	for i< len(s) && j< len(p) {
+		if marker, ok := inds[j]; ok {
+			value := ""
+			for i<len(s) && (j+len(marker) == len(p) || s[i] != p[j+len(marker)]) {
+				value += string(s[i])
+				i++
+			}
+			// if marker is repeated in the pattern, make sure that the corresponding values
+			// are same and throw error if not.
+			if prevValue, ok := m[marker]; ok && prevValue != value {
+				return nil, errors.New("Same marker is found to have different values in field value.")
+			}
+			m[marker] = value
+			j+=len(marker)
+		} else {
+			if s[i] != p[j] {
+				return nil, errors.New("Unable to derive values for markers. Please create setters for all markers.")
+			}
+			i++
+			j++
+		}
+	}
+	if i<len(s) || j<len(p) {
+		return nil, errors.New("Unable to derive values for markers. Please create setters for all markers.")
+	}
+	return m, nil
+}
+
+// GetStartIndices returns the start indices of all the markers in the pattern
+func (c SubstitutionCreator) GetStartIndices() (map[int]string, error) {
+	inds := make(map[int]string)
+	p := c.Pattern
+	for _, value := range c.Values {
+		m := value.Marker
+		found := false
+		for i, _ := range p {
+			if p[i:int(math.Min(float64(i+len(m)), float64(len(p))))] == m {
+				inds[i] = m
+				found = true
+			}
+		}
+		if !found {
+			return nil, errors.New("Unable to find marker " + m + " in the pattern")
+		}
+	}
+	return inds, nil
```

Comment:
check for overlapping values in inds -- e.g. if there is another map key in the range of i and i + len(inds[i]) then there is an error

Reply:
Is this the case where one marker is substring of other? The algorithm handles it but throws generic error. There will be many such cases and question is to what level we should report specific errors. Algorithm is more stringent and looks for perfect match of pattern and field value. Can you quote an example and I can verify.

Reply:
Please refer to unit test TestGetValuesForMarkersSubStngMarkers